### PR TITLE
Refactor histogram to use map blocks

### PR DESCRIPTION
**Pull request by rabernat**
As discussed in https://github.com/ocean-transport/coiled_collaboration/issues/8, I feel like there must be a more efficient implementation of xhistogram's core logic in terms of handling dask arrays. A central problem with our implementation is the use of reshape on dask arrays. This causes inefficient chunk manipulations (see https://github.com/dask/dask/issues/5544).

We should be able to just apply the bin_count function to every block and then sum over all blocks to get the total bin count. If we go this route, we will no longer use dask.array.histogram at all. This should result in much simpler dask graphs.

I have not actual figured out how to solve the problem, but this PR gets started in that direction. The challenge is in determining the arguments to pass to `map_blocks`.

### CODE REVIEW

**Comment by dcherian** [comment]
:+1: You know the number of bins or groups so you can "tree reduce" it without any reshaping as you say. The current version of dask's bincount does exactly that:
https://github.com/dask/dask/blob/98a65849a6ab2197ac9e24838d1effe701a441dc/dask/array/routines.py#L677-L698

It's what inspired the core of my groupby experiment: https://github.com/dcherian/dask_groupby/blob/373a4dda4d618848b5d02cca551893fee36df092/dask_groupby/core.py#L361-L368 which only reshapes if you know absolutely nothing about the "expected" groups.

IIUC your proposal is implemented in tensordot as a blockwise-which-adds-dummy-axis followed by sum. 
https://github.com/dask/dask/blob/98a65849a6ab2197ac9e24838d1effe701a441dc/dask/array/routines.py#L295-L310

I found blockwise a lot easier to handle than map_blocks for this stuff

**Comment by jrbourbeau** [comment]
cc @gjoseph92

**Comment by rabernat** [comment]
@dcherian - at this point it is clear that you know a lot more about this than me. Would you have any interest in taking a crack at implementing the approach you just described? I think this PR sets it up nicely so all you have to do is plug in the tree reduce code.

File: xhistogram/core.py [hunk 1]
```diff
@@ -222,54 +281,39 @@ def histogram(
...
+            bins,
+            density,
+            drop_axis=axis,
+            new_axis="what?",
+            chunks="what?",
+            meta=np.array((), dtype=np.int64),
+        )
+        # sum over the block dims
+        block_dims = "?"
+        bin_counts = bin_counts.sum(block_dims)
```
[the remark this answers is not in the document]
Comment (dcherian) [reply xhistogram/core.py:314]: Well I tried to start looking at it but nothing works without #48 being fixed

[the remark this answers is not in the document]
Comment (jrbourbeau) [reply xhistogram/core.py:314]: There's an upstream fix in `dask` (xref https://github.com/dask/dask/pull/7594) which should be merged soon

**Comment by gjoseph92** [comment]
Okay, I _think_ this is the skeleton for how we could implement it: https://gist.github.com/gjoseph92/1ceaf28937f485012bc1540eb2e8da7d

We just need the actual `bincount`-ish function to be applied over the chunk pairs. And to handle weights, and the bins being dask arrays. But I think those are "just" a matter of broadcasting and lining up the dimension names to what I have here.

Now that I'm seeing how much cleaner the graphs can be in the n-dimensional chunking case (I think the examples I'd looked at earlier all had one chunk along the aggregated dimensions), I fully agree with @rabernat that blockwise could improve the graph structure a lot.

**Comment by dcherian** [comment]
:+1: That's exactly it AFAICT. It's the same trick as tensordot and isin.

**Comment by rabernat** [comment]
@gjoseph92 that looks amazing. Exactly what I had in mind originally with this PR but didn't have the dask chops to actually implement. Let me know how I can help move this forward.

**Comment by TomNicholas** [comment]
I'm happy to have a go at implementing this too. I would eventually like to integrate it upstream in xarray, but only once we've got the best dask function for the job working.

One question - what would the numpy equivalent of @gjoseph92 's solution be? Is it just to apply `bincount` once over the whole array without calling `blockwise`?

**Comment by dougiesquire** [comment]
> I'm happy to have a go at implementing this too. I would eventually like to integrate it upstream in xarray, but only once we've got the best dask function for the job working.

@TomNicholas, for someone that's not up with the lingo, do you mean move histogram functionality into xarray?

> One question - what would the numpy equivalent of @gjoseph92 's solution be? Is it just to apply `bincount` once over the whole array without calling `blockwise`?

For 1D arrays at least, I think numpy histogram actually loops over blocks and cumulatively sums the `bincount` applied to each block, with [a note that this is both faster and more memory efficient for large arrays](https://github.com/numpy/numpy/blob/v1.20.0/numpy/lib/histograms.py#L824).

**Comment by TomNicholas** [comment]
> do you mean move histogram functionality into xarray

Yes exactly. It's arguably within the scope of xarray (which already has a `.hist` method, it just currently reduces over all dimensions), and has functions for `groupby` operations and similar types of reshaping already.

In the long term the advantages would be:
- Domain-agnostic functionality would be living in a domain-agnostic library
- More users would use the code & benefit from it
- More users means more chance of finding bugs, and more help with extending to add new features
- Not having the maintenance burden of a separate package

> I think numpy histogram actually loops over blocks

Oh that's interesting, thanks!

**Comment by dougiesquire** [comment]
> Yes exactly. It's arguably within the scope of xarray

I think this is a terrific idea. Happy to help out however I can

**Comment by TomNicholas** [comment]
> I think this is a terrific idea. Happy to help out however I can

Great! Well anyone can do the PR(s) to xarray - I would be happy to provide guidance & code reviews over there if you wanted to try it?

The steps as I see it are:
- Get the dask.blockwise bare version working efficiently, probably here (could build off Ryan's PR in another PR if we want)
- Discuss the external API we want in a "feature proposal" issue on xarray (I'll open one for that with some ideas)
- Copy over whatever tests + docs are transferable to an xarray PR
- Get that minimum version working in the xarray PR
- Once the minimum version works in xarray I would probably only worry about adding weighted functionality and so on in a later PR
- Eventually deprecate xhistogram
- profit

**Comment by dougiesquire** [comment]
Yup I think it's a good idea to get the dask.blockwise implementation working well here as a first step :thumbsup:

**Comment by TomNicholas** [comment]
So looking at the code in `xhistogram.core`, I have some questions:

To implement @gjoseph92 's solution (which looks great) then we would need to write a new `bincount` function that matches the signature described in his notebook. The `_bincount` function we currently have has a different signature (though I'm not sure exactly what signature it has because it's an internal function so doesn't have a docstring), so would need to be significantly changed. Then `_bincount` calls these other functions like `_bincount_2d_vectorized`, which I think is specific to the `reshape` dask algorithm we are currently using? If we are changing all these functions are we essentially just rewriting the whole of `xhistogram.core`? I'm not saying we shouldn't do that, I'm just trying to better understand what's required. Can any of this be copied over from other implementations (e.g. dask.array.histogramdd)?

**Comment by gjoseph92** [comment]
> If we are changing all these functions are we essentially just rewriting the whole of `xhistogram.core`

Yes, I think we sort of are. xhistogram's current approach is dask-agnostic (ish, besides `_dispatch_bincount`)—it uses a bunch of NumPy functions and lets dask parallelize them implicitly. With the `blockwise` approach, we'd sort of have a shared "inner kernel" (the core bincount operation), then slightly different code paths for mapping that over a NumPy array versus explicitly parallelizing it over a dask array.

The difference is very small: for a NumPy array, apply the core bincount function directly to the array. For a dask array, use `blockwise` to apply it. (You can think of the NumPy case as just a 1-chunk dask array.) Then in both cases, finish with `bincounts.sum(axis=axis)`.

So in a way, the bincount "inner kernel" is all of the current xhistogram core, just simplified to not have to worry about dask arrays, and to return a slightly different shape. I think this change might actually make xhistogram a bit easier to read/reason about. Additionally, I think we should consider writing this bincount part in numba or Cython eventually—not just because it's faster, but because being able to just write a `for` loop would be way easier to read than the current `reshape` and `ravel_multi_index` logic, which is hard to wrap your head around.

Also, it might be worth passing NumPy arrays of sufficient size (or any size) into `dask.array.from_array`, putting them through the dask logic, and computing the result with `scheduler="threads"`. This would be a quick and easy way to get the loop-over-blocks behavior @dougiesquire [was describing](https://github.com/xgcm/xhistogram/pull/49#issuecomment-833908104), (maybe) a lower peak memory footprint, plus some parallelism.

**Comment by TomNicholas** [comment]
That's very helpful, thanks @gjoseph92 .

> I think this change might actually make xhistogram a bit easier to read/reason about.

I agree.

> Additionally, I think we should consider writing this bincount part in numba or Cython eventually

That would be cool, but it would also affect whether or not we could integrate the functionality upstream in xarray, so I suggest that we get a version working with just dask initially so we can see if it's fast "enough"?

> computing the result with scheduler="threads"

That might even be like a "best of both" solution, ish?

**Comment by dougiesquire** [comment]
Thanks @gjoseph92 for that very helpful description! 

@TomNicholas I'm going on leave next week and I'm hoping to work on this a little during that. My plan is to spend some time understanding/implementing the `blockwise` approach and then to open another PR once things are somewhat functional. But I want to make sure I don't tread on your toes or duplicate your efforts. What's the best way to align our efforts?

**Comment by TomNicholas** [comment]
> I'm hoping to work on this a little during that

Great!

> But I want to make sure I don't tread on your toes or duplicate your efforts.

I wouldn't worry about that too much. I think we will both want to understand the resultant code, so time spent understanding it is always well-spent. For development we should just both post whatever we try out - as a gist, a PR, or could even start a new dev branch on this repo.

File: xhistogram/core.py [hunk 2]
```diff
@@ -146,6 +147,64 @@ def _histogram_2d_vectorized(
...
+    all_arrays_reshaped = [reshape_input(a) for a in all_arrays_broadcast]
+
+    if weights_broadcast is not None:
+        weights_reshaped = reshape_input(weights_broadcast)
+    else:
+        weights_reshaped = None
+
+    bin_counts = _bincount_2d_vectorized(
+        *all_arrays_reshaped, bins=bins, weights=weights_reshaped, density=density
+    )
```
Comment (rabernat) [inline xhistogram/core.py:196]: This is not obvious but: bin_counts is not necessarily 2D. Its ndim is 1 + len(all_arrays). So it represents and ND historgram.

**Comment by rabernat** [comment]
In my chat today with @TomNicholas, we realized this PR was already quite close to working. I just needed to understand how to use blockwise (rather than map_blocks). Tests are all green. I have not examined the dask graphs yet, but I think they will be improved.

Regardless of whether this works, Tom raised the point that the current implementation is quite opaque. There are many nested layers of internal functions that make the logic hard to follow. So it might not be a bad idea to just refactor everything anyway.

File: xhistogram/core.py [hunk 3]
```diff
@@ -222,54 +278,68 @@ def histogram(
...
+        for arg in all_arrays:
+            blockwise_args.append(arg)
+            blockwise_args.append(input_index)
+
+        bin_counts = dsa.blockwise(
+            _bincount,
+            out_index,
+            *blockwise_args,
+            new_axes=new_axes,
+            meta=np.array((), dtype) ** bincount_kwargs,
```
Comment (TomNicholas) [inline xhistogram/core.py:336]: @rabernat this is a syntax bug, there is a missing comma between arguments. It should be 

```python
meta=np.array((), dtype), 
**bincount_kwargs,
```

This actually causes this blockwise algorithm to error for chunked inputs, because the bins never get passed in. The reason the tests still pass is because the lack of tests with explicit chunking means this code path is not really tested in most of the current tests.

I've made some hypothesis tests to cover different types of chunking but I'm currently trying to fix some other bugs they revealed.

Comment (rabernat) [inline xhistogram/core.py:336]:
```suggestion
            meta=np.array((), dtype), 
            **bincount_kwargs,
```

**Comment by rabernat** [comment]
@TomNicholas - today on our call you mentioned that this PR doesn't actually work with Dask. 🤣  That was absolutely true of 9f20e95. However, I think I have updated things now so that they do actually work. 

Here is an example.

```python
import numpy as np
import dask.array as dsa
import xhistogram.core as xhc

csize = 80
nz = 20
ntime = 2
fac = 4

bins = np.arange(10)

data = dsa.ones((nz, fac * csize, fac * csize),
                chunks=(nz, csize, csize))
display(data)
hist = xhc.histogram(data, bins=[bins], axis=[0, 1])
hist_eager = xhc.histogram(data.compute(), bins=[bins], axis=[0, 1])
display(hist)
display(hist.visualize())
np.testing.assert_allclose(hist.compute(), hist_eager)
```

image

image

Regardless, the fact that 9f20e95 did not raise any test failures reveals a major gap in our test coverage. You mentioned you were working on expanding test coverage of dask cases. It would be great to do that in a separate PR, merge it, and then rebase this on top of it to see whether the implementation here passes the new tests.

**Comment by TomNicholas** [comment]
> now they do actually work

@rabernat nice! Thanks for fixing that. Also that task graph looks beautiful! Is it still as neat for a multi-variable histogram?

I was really struggling to work out what was the problem with your PR, and I still don't really understand how you fixed it - IMO that means the code should still be refactored to be clearer, which I will have a go at.

> You mentioned you were working on expanding test coverage of dask cases

Yes it's pretty surprising that broken code still passed the tests... but I've added dask tests over in #57 now.

**Comment by TomNicholas** [comment]
> Is it still as neat for a multi-variable histogram?

For anyone wondering yes it is:

image

**Comment by jbusecke** [comment]
> For anyone wondering, yes it is (for same but with two data variables)

Amazing! Is this also including weights?

**Comment by TomNicholas** [comment]
With weights you get this:

```python
csize = 80
nz = 20
ntime = 2
fac = 4

bins1 = np.arange(10)
bins2 = np.arange(8)

data1 = dsa.random.random((nz, fac * csize, fac * csize),
                          chunks=(nz, csize, csize))
data2 = dsa.random.random((nz, fac * csize, fac * csize),
                          chunks=(nz, csize, csize))
weights = dsa.random.random((nz, fac * csize, fac * csize),
                          chunks=(nz, csize, csize))

display(data)
hist = xhc.histogram(data1, data2, bins=[bins1, bins2], weights=weights, axis=[0, 1])
hist_eager = xhc.histogram(data1.compute(), data2.compute(), bins=[bins1, bins2], weights=weights.compute(), axis=[0, 1])
display(hist)
display(hist.visualize())
np.testing.assert_allclose(hist.compute(), hist_eager)
```

image

**Comment by jbusecke** [comment]
Just beautiful! I suspect this will help tremendously over at https://github.com/ocean-transport/coiled_collaboration/issues/9. 

Do you think this is ready to test under real-world conditions?

**Comment by TomNicholas** [comment]
> Do you think this is ready to test under real-world conditions?

You could try it, but probably not, because chunking has no explicit tests, apart from the ones I added in #57, which do not all yet pass :confused:

**Comment by TomNicholas** [comment]
@jbusecke scratch that - it looks like it was my tests that were the problem after all - should be green to try this out now!

**Comment by jbusecke** [comment]
Whoa! This works really well!

Dask stream from a synthetic example, which would fail often (https://github.com/ocean-transport/coiled_collaboration/issues/9)
image

I ran this on the google deployment with the 'hacky pip install' described [here](https://pangeo.io/cloud.html#dask-software-environment). Will try it out for realistic CMIP6 data next.

I can't thank you enough for implementing this! I know who I will thank at the next CDSLab meeting!

**Comment by jbusecke** [comment]
Ok so after some tuning of the chunksize I got some pretty darn satisfying results for the real world example:
image

This crunched through a single member in under 2 minutes with 20 workers. Absolutely no spilling!

File: xhistogram/core.py [hunk 4]
```diff
@@ -222,54 +274,77 @@ def histogram(
...
-        weights_reshaped = None
+    bincount_kwargs = dict(weights=has_weights, axis=axis, bins=bins, density=density)
 
-    n_inputs = len(all_args_reshaped)
-    bins = _ensure_bins_is_a_list_of_arrays(bins, n_inputs)
+    # here I am assuming all the arrays have the same shape
+    # probably needs to be generalized
+    input_indexes = [tuple(range(a.ndim)) for a in all_arrays]
+    input_index = input_indexes[0]
+    assert all([ii == input_index for ii in input_indexes])
```
Comment (gjoseph92) [inline xhistogram/core.py:298]: nit: it seems this is only needed in the `if _any_dask_array` case; maybe move it into that conditional?

> here I am assuming all the arrays have the same shape

Maybe you'd want to `np.broadcast_arrays(*all_arrays)` here (note that this will dispatch to `da.broadcast_arrays` automatically thanks to NEP-18)?

Alternatively, I think you could get away without broadcasting with some slight modifications to the blockwise indices. Specifically, rather than using `input_index` for all arrays, use each array's own indices. Then blockwise will basically express any the broadcasting for you in the graph:

```python
input_indexes = [tuple(range(a.ndim)) for a in all_arrays]
largest_index = max(input_indexes, key=len)
if axis is not None:
    drop_axes = tuple([ii for ii in largest_index if ii in axis])
else:
    drop_axes = largest_index

out_index = largest_index + tuple(new_axes)
...
blockwise_args = []
for arg, index in zip(all_arrays, input_indexes):
    blockwise_args.append(arg)
    blockwise_args.append(index)
```

File: xhistogram/core.py [hunk 5]
```diff
@@ -146,6 +147,57 @@ def _histogram_2d_vectorized(
     return bin_counts
 
 
+def _bincount(*all_arrays, weights=False, axis=None, bins=None, density=None):
+
+    # is this necessary?
+    all_arrays_broadcast = broadcast_arrays(*all_arrays)
```
Comment (gjoseph92) [inline xhistogram/core.py:153]: If you go with the change suggested in https://github.com/xgcm/xhistogram/pull/49#discussion_r640872588 (no `np.broadcast` in top-level `histogram`), then I think this broadcast _is_ necessary. Otherwise, if `a0` happens to have dimensions that would be added or expanded by broadcasting, then your `kept_axes_shape` calculation could be wrong, and `reshape_input` might move the wrong axes.

However, if you added an `np.broadcast` in top-level `histogram`, then I think you could remove this one here.

Comment (rabernat) [reply xhistogram/core.py:153]: So with this PR, none of this code path will ever see dask arrays. So we can probably remove the duck_array_ops module.

**Comment by TomNicholas** [comment]
So now that we've successfully chunked over all dims, do we still need any of the "block size" logic? The main benefit of being able to specify `block_size` is a lowered memory footprint right? But if we can now chunk over any dim then the dask chunks can always be made smaller for a given problem, and the user can always lower their memory footprint by using smaller chunks instead?

**Comment by rabernat** [comment]
> Or do we still want it for the same reasons that [numpy.histogram has it](https://github.com/numpy/numpy/blob/v1.20.0/numpy/lib/histograms.py#L824)?

My understanding was that this is a low-level optimization, aimed at making the core bincount routine faster. Dask is responsible for scaling out, but having a faster low-level routine is always desireable as well. That said, I have not benchmarked anything to verify it actually makes a difference.

File: xhistogram/core.py [hunk 6]
```diff
@@ -332,17 +378,59 @@ def reshape_input(a):
             )
     else:
         bins = [
-            np.histogram_bin_edges(a, b, r, weights_reshaped)
-            for a, b, r in zip(all_args_reshaped, bins, range)
+            np.histogram_bin_edges(a, b, r, weights)
+            for a, b, r in zip(all_arrays, bins, range)
```
Comment (rabernat) [inline xhistogram/core.py:382]: I brought this code in when I merged with upstream/master and picked up the new bin edges logic introduced by @dougiesquire in #45. Unfortunately the tests are going to fail because of shape mismatches between arrays and weights. Due to the refactor in this PR, the high-level `histogram` function no longer has access to the reshaped / broadcasted arrays. So we will need some new solution here.

Comment (dougiesquire) [reply xhistogram/core.py:382]: Maybe it's cleanest/easiest then to `np.broadcast_arrays(*all_arrays)` early on as @gjoseph92 suggests in https://github.com/xgcm/xhistogram/pull/49#discussion_r640872588? Then we also don't need to worry about each array potentially having a different `input_index`.

Another option I guess would be to do the broadcasting in xarray.histogram and have core.histogram expect that all `args` and `weights` are the same shape (somewhat replicating what numpy.histogram expects).

File: xhistogram/core.py (hunk 4 above)
Comment (TomNicholas) [reply xhistogram/core.py]: @gjoseph92 is there any performance reason to prefer making blockwise handle the broadcasting? It seems simplest to just broadcast everything at the start, but does that mean carrying around larger arrays than necessary?

Comment (gjoseph92) [reply xhistogram/core.py]: The arrays shouldn't actually increase in physical size; their shape will just be expanded. The only advantage of using blockwise alone is that the graph will be a little smaller:
* No extra reshape layers added prior to the blockwise (which are also fully materialized layers and slightly more expensive to serialize to the scheduler)
* The low-level graphs of the arrays input into blockwise won't contain superfluous keys for the extra broadcast dimensions

In practice, I very much doubt it would be an appreciable difference, but not having compared them I can't say for sure. There is also a slight optimization advantage with pure-blockwise when there is exactly 1 input array, and it's already the result of a blockwise operation.

File: xhistogram/core.py (hunk 6 above)
Comment (TomNicholas) [reply xhistogram/core.py:382]: @rabernat it - the docstring of `np.histogram_bin_edges` says that the weights argument "is currently not used by any of the bin estimators, but may be in the future." So it currently doesn't do anything: you can make all the tests pass again simply by deleting the weights argument from the call to `np.histogram_bin_edges`.

Separately I do think that we should consider broadcasting early on to make the indexing easier to follow, but it would be nice to get this PR merged and discuss that refactoring elsewhere.

File: xhistogram/core.py [hunk 7]
```diff
@@ -332,17 +378,59 @@ def reshape_input(a):
             )
     else:
         bins = [
-            np.histogram_bin_edges(a, b, r, weights_reshaped)
-            for a, b, r in zip(all_args_reshaped, bins, range)
+            np.histogram_bin_edges(a, b, r, weights)
```
Comment (TomNicholas) [inline xhistogram/core.py:381]:
```suggestion
            np.histogram_bin_edges(a, b, r)
```

File: xhistogram/core.py (hunk 4 above)
Comment (TomNicholas) [reply xhistogram/core.py]: That's interesting, thanks.

However it's just occurred to me that surely relying on blockwise to do all the broadcasting is not an option, simply because the non-dask case doesn't use blockwise. Supporting a numpy-only code path (as we would likely want both for debugging and for fitting xarray's standard practice) means broadcasting all arrays before they are fed to `_bincount`.

**Comment by rabernat** [comment]
Tom, the only thing that failed after I accepted your suggestion was the linting. So please go ahead and merge this if you think it's ready! Lots of stuff is blocking on this PR.

**Comment by jbusecke** [comment]
🚀
